lib/resty/core/socket.lua

```diff
@@ -0,0 +1,124 @@
+local base = require "resty.core.base"
+local debug = require 'debug'
+local ffi = require 'ffi'
+
+
+local error = error
+local tonumber = tonumber
+local registry = debug.getregistry()
+local ffi_new = ffi.new
+local ffi_string = ffi.string
+local C = ffi.C
+local get_string_buf = base.get_string_buf
+local get_size_ptr = base.get_size_ptr
+local tostring = tostring
+
+
+local option_index = {
+    ["keepalive"]   = 1,
+    ["reuseaddr"]   = 2,
+    ["tcp-nodelay"] = 3,
+    ["sndbuf"]      = 4,
+    ["rcvbuf"]      = 5,
+}
+
+
+ffi.cdef[[
+typedef struct ngx_http_lua_socket_tcp_upstream_s
+    ngx_http_lua_socket_tcp_upstream_t;
+
+int
+ngx_http_lua_ffi_socket_tcp_getoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    int opt, int *val, unsigned char *err, size_t *errlen);
+
+int
+ngx_http_lua_ffi_socket_tcp_setoption(ngx_http_lua_socket_tcp_upstream_t *u,
+    int opt, int val, unsigned char *err, size_t *errlen);
+]]
+
+
+-- int is safe here since the maximum is 2^31-1
+local output_value_buf = ffi_new("int[1]")
+local FFI_OK = base.FFI_OK
+local SOCKET_CTX_INDEX = 1
+local ERR_BUF_SIZE = 128
+
+
+local function get_tcp_socket(cosocket)
+    local tcp_socket = cosocket[SOCKET_CTX_INDEX]
+    if not tcp_socket then
+        error("bad tcp socket")
```

Comment:
This error is confusing. We could do better. What does this really mean? The socket is never created? Never connected? Errored? What is it?

Reply:
accept, "bad tcp socket" => "socket is never created nor connected"

Reply:
“nor” => "or"
"never" => "not"